src/main/kotlin/org/rust/lang/core/types/infer/Expressions.kt

```diff
@@ -127,10 +127,38 @@ fun inferExpressionType(expr: RsExpr): RustType {
                 RustUnknownType
         }
 
+        is RsArrayExpr -> inferArrayType(expr)
+
         else -> RustUnknownType
     }
 }
 
+private fun inferArrayType(expr: RsArrayExpr): RustType {
+    val (elementType, size) = if (expr.semicolon != null) {
+        if (expr.exprList.size != 2) return RustUnknownType
+        val (elementExpr, sizeExpr) = expr.exprList
+        val size = calculateArraySize(sizeExpr) ?: return RustUnknownType
+        Pair(elementExpr.type, size)
+    } else {
+        if (expr.exprList.isEmpty()) return RustUnknownType
+        Pair(expr.exprList[0].type, expr.exprList.size)
+    }
+    return RustArrayType(elementType, size)
+}
+
+// TODO: support constants and compile time expressions
+private fun calculateArraySize(expr: RsExpr): Int? {
```

Comment:
Yes, it's a good idea. I've made it public, moved into `org.rust.lang.core.psi.ext.RsArrayType.Kt`(I think it's better place for this function if we want to use it for size calculation of `RsArrayType`) and use from both places.